src/schema/document/de.rs

```diff
@@ -0,0 +1,1023 @@
+    fn visit_bool(&self, _val: bool) -> Result<Self::Value, DeserializeError> {
+        Err(DeserializeError::UnsupportedType(ValueType::Bool))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits a datetime value.
+    fn visit_datetime(&self, _val: DateTime) -> Result<Self::Value, DeserializeError> {
+        Err(DeserializeError::UnsupportedType(ValueType::DateTime))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits an IP address value.
+    fn visit_ip_address(&self, _val: Ipv6Addr) -> Result<Self::Value, DeserializeError> {
+        Err(DeserializeError::UnsupportedType(ValueType::IpAddr))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits a facet value.
+    fn visit_facet(&self, _val: Facet) -> Result<Self::Value, DeserializeError> {
+        Err(DeserializeError::UnsupportedType(ValueType::Facet))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits a bytes value.
+    fn visit_bytes(&self, _val: Vec<u8>) -> Result<Self::Value, DeserializeError> {
+        Err(DeserializeError::UnsupportedType(ValueType::Bytes))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits a pre-tokenized string value.
+    fn visit_pre_tokenized_string(
+        &self,
+        _val: PreTokenizedString,
+    ) -> Result<Self::Value, DeserializeError> {
+        Err(DeserializeError::UnsupportedType(ValueType::PreTokStr))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits an array.
+    fn visit_array<'de, A>(&self, _access: A) -> Result<Self::Value, DeserializeError>
+    where A: ArrayAccess<'de> {
+        Err(DeserializeError::UnsupportedType(ValueType::Array))
+    }
+
+    #[inline]
+    /// Called when the deserializer visits a object value.
+    fn visit_object<'de, A>(&self, _access: A) -> Result<Self::Value, DeserializeError>
+    where A: ObjectAccess<'de> {
+        Err(DeserializeError::UnsupportedType(ValueType::Object))
+    }
+}
+
+/// Access to a sequence of values which can be deserialized.
+pub trait ArrayAccess<'de> {
+    /// A indicator as to how many values are in the object.
+    ///
+    /// This can be used to pre-allocate entries but should not
+    /// be depended on as a fixed size.
+    fn size_hint(&self) -> usize;
+
+    /// Attempts to deserialize the next element in the sequence.
+    fn next_element<V: ValueDeserialize>(&mut self) -> Result<Option<V>, DeserializeError>;
+}
+
+/// TODO: Improve docs
+pub trait ObjectAccess<'de> {
+    /// A indicator as to how many values are in the object.
+    ///
+    /// This can be used to pre-allocate entries but should not
+    /// be depended on as a fixed size.
+    fn size_hint(&self) -> usize;
+
+    /// Attempts to deserialize the next key-value pair in the object.
+    fn next_entry<V: ValueDeserialize>(&mut self) -> Result<Option<(String, V)>, DeserializeError>;
+}
+
+/// TODO: Improve docs
+pub struct GenericDocumentDeserializer<'de, R> {
```

Comment:
Why is this called generic? On the format side it seems to assume Vint and BinarySerializable

Reply:
By generic I meant in the literal sense, i.e. it's generic over a given reader. Probably can be renamed to something better.